flang/include/flang/Lower/Bridge.h

```diff
@@ -60,7 +60,10 @@ class LoweringBridge {
   //===--------------------------------------------------------------------===//
 
   mlir::MLIRContext &getMLIRContext() { return context; }
-  mlir::ModuleOp &getModule() { return *module.get(); }
+  mlir::ModuleOp &getModule() {
+    assert(module && "This bridge is missing an MLIR module");
```

Comment:
This seems like it is maybe crossing the line into paranoia here. Nothing is going to go correctly without a Module. And if one somehow gets here before creating the module, the compiler is going to exit immediately anyway.

We should be mindful not to over-assert. I've seen projects where coders added assertions basically everywhere to a point where performance got bad enough, people disabled the assertions to do their work. Which is exactly what we don't want. :)

Reply:
> Nothing is going to go correctly without a Module.

Probably not. But if, for whatever reason, there is no Module here, then this assert will point people in the right direction. And I couldn't find anything better that would indicate whether the lowering went fine (or not). [LoweringBridge::lower](https://github.com/llvm/llvm-project/blob/main/flang/include/flang/Lower/Bridge.h#L98-L99) simply returns `void`.

> We should be mindful not to over-assert.

LLVM suggests that one should [Assert Liberally](https://llvm.org/docs/CodingStandards.html#id44). Also:
```bash
$ grep -R " assert(" llvm/ | wc -l
25024
$ grep -R " assert(" clang/ | wc -l
10490
$ grep -R " assert(" mlir/ | wc -l
2191
$ grep -R " assert(" flang/ | wc -l
187
```
So, one additional `assert` in Flang is unlikely to have a huge impact on performance here.

Reply:
That is misconstruing what I meant here.

```c
   for (i = 0; i < 10000000; i++) {
     assert(TestSomeConditionPossiblyExpensively() && "only if assertions were enabled");
     ...
   }
``` 
is not "one additional `assert`" in terms of runtime performance.

The test tools give tracebacks so it is quite trivial to pinpoint an egregious and fundamental error in the code like this one is adding an explicit test for.

Reply:
Backing up and starting at the beginning.

This assertion is guarding against the _use_ (reference to) of the ModuleOp as being non-null. Let's break that down.

First, the ModuleOp is the container into which the bridge (the code we're changing here) inserts _all_ of the IR. The ModuleOp can never be null and the bridge also correctly produce IR. That will never happen in a correct program. This is very similar to using a file. One can write and compile a C program that writes to a file before opening it, but that program will crash or produce garbage results.

Second, the assert is guarding a reference to the ModuleOp. A running executable may have many references to the ModuleOp, some obvious and some not. We can call this the set _R_ = { _r1_ , _r2_ , ... _rn_ } of references. The assert guarantees that for some subset _R'_ = { _ri_ , _rj_ , ... _rk_ } of references to the ModuleOp, that the ModuleOp is (still) non-null. Returning to the file analogy, it is like checking that the file is still open after some arbitrary number of reads/writes to the file have been performed.

I fully expect that the cardinality of _R'_ is small so this test is not a huge runtime burden here. But the idea of checking on use can lead one to writing over-asserting code. For example, one could write code like the following everywhere in the bridge:
```c++
mlir::Value createFoo(mlir::Value arg1, mlir::Value arg2, mlir::Value arg3, mlir::Value arg4) {
  assert(arg1 && "argument value cannot be nullptr");
  assert(arg2 && "argument value cannot be nullptr");
  assert(arg3 && "argument value cannot be nullptr");
  assert(arg4 && "argument value cannot be nullptr");
  return builder.create<Foo>(loc, type, arg1, arg2, arg3, arg4);
}
```

This liberally uses assert before using the four arguments before we build a `Foo`. It compiles, etc.

No one writes code like this however. One expects whatever constructed the ssa-values to have made them correctly and not to have failed. Furthermore, if some of them are nullptr, the verifier will fail if some other crash doesn't happen. And those failures and crashes tell us that something is really wrong with this code without the assertion overhead of testing every ssa-value and operation everywhere before using it.

Reply:
> That is misconstruing what I meant here.

That wasn't my intention. 

> The ModuleOp can never be null and the bridge also correctly produce IR.

That's the case _at the moment_. But there is no guarantee that this won't change in the future (with the number of contributors growing, it might be hard to make sure that no such code is introduced). Also, it's an assumption that's not verified anywhere apart from where the Lowering Bridge is [created](https://github.com/llvm/llvm-project/blob/main/flang/lib/Lower/Bridge.cpp#L842). There's [validModule](https://github.com/llvm/llvm-project/blob/a87d3ba61c64bf957f2f9369e4d076cbbf3d486f/flang/include/flang/Lower/Bridge.h#L87), but it will seg-fault when the module is not valid (proposed fix: https://reviews.llvm.org/D119705).

As for run-time-costs, asserts are disabled in release builds with `-DNDEBUG`. Also, there are ~37k `assert`s in sub-projects that Flang depends on and _only_ ~180 `assert`s in Flang. The cost of this particular `assert` is going to be negligible. If people decide to disable `assert`s in their local builds, we have buildbots upstream that have them enabled (so they will be tested). We recommend that everyone enables assertions when building LLVM and I don't recall that being a run-time issue for anyone.

I feel that this has evolved into a discussion on the merits of using `assert`s in general. IIUC, this part of Flang follows LLVM's coding style, which is clear on the usage of `assert`s. Also, this change was already reviewed and accepted upstream. I'm just making sure that `fir-dev` stays in sync with `main`. And in the process of upstreaming, we have been adding a lot of tests, `TODO`s, `assert`s etc. This is a similar change.

Reply:
@banach-space You are being too literal wrt `assert`.   Flang has 2200 calls to CHECK.

@schweitzpgi > No one writes code like this however.
Everyone who learned to code from Dr. Z writes code like that!

Reply:
The creation of the ModuleOp is already being asserted at its creation point.
https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/lib/Lower/Bridge.cpp#L3051-L3053

This assertion is not needed.